### PR TITLE
Version conflict

### PR DESCRIPTION
**Pull request by lielShraga**

### CODE REVIEW

File: server/cmwell-fts/src/main/scala/cmwell/fts/FTSService.scala [hunk 1]
```diff
@@ -504,12 +504,17 @@ class FTSService(config: Config) extends NsSplitter{
...
-            bulkItemResponse.getFailureMessage.startsWith("VersionConflictEngineException")
+          val (versionConflictErrors, unexpectedErrors) = nonRecoverableFailures.partition {
+            case (itemResponse, _) =>
+              itemResponse.getFailureMessage.contains("VersionConflictEngineException")
           }
 
           // log errors
+          versionConflictErrors.foreach{ case (itemResponse, esIndexRequest) =>
+            val infotonPath = infotonPathFromDocWriteRequest(esIndexRequest.esAction)
+            logger.error(s"ElasticSearch non recoverable version conflict failure on doc id:${itemResponse.getId}, path: $infotonPath . " +
```
Comment (e-orz) [inline server/cmwell-fts/src/main/scala/cmwell/fts/FTSService.scala]: should be infto

File: server/cmwell-fts/src/main/scala/cmwell/fts/FTSService.scala [hunk 2]
```diff
@@ -540,7 +545,22 @@ class FTSService(config: Config) extends NsSplitter{
               logger.error(s"exhausted all retries attempts. logging failures to RED_LOG and returning results ")
               promise.success(SuccessfulBulkIndexResult(indexRequests, bulkResponse))
             }
-          } else {
+          } else if(versionConflictErrors.length > 0){
```
Comment (e-orz) [inline server/cmwell-fts/src/main/scala/cmwell/fts/FTSService.scala]: the code should also handle the case of some recoverable and some not recoverable errors

File: server/cmwell-fts/src/main/scala/cmwell/fts/FTSService.scala [hunk 3]
```diff
@@ -540,7 +545,22 @@ class FTSService(config: Config) extends NsSplitter{
...
+                  case Some(ti) => Success(SuccessfulIndexResult(sir._1.getId, Option(ti._1.indexTime)))
+                  case None => Success(SuccessfulIndexResult(sir._1.getId, None))
+                }
+                case Failure(e) => logger.error("Got exception from es while get index time for version conflict, error=" +
+                  e.getMessage + ",uuid=" + sir._1.getId)
+                  Failure(e)
+              }
+            }
+            val allSucessIndexes = cmwell.util.concurrent.successes(bulks.toList)
+            allSucessIndexes.map {x=> promise.success(bulkSuccessfulResult ++  SuccessfulBulkIndexResult(x))}
```
Comment (e-orz) [inline server/cmwell-fts/src/main/scala/cmwell/fts/FTSService.scala]: You should always return all the results. The succesfull and the failed ones.

Comment (e-orz) [inline server/cmwell-fts/src/main/scala/cmwell/fts/FTSService.scala]: `SuccessfulBulkIndexResult(x)` actually wraps the result with redundant `SuccessfulBulkIndexResult`

File: server/cmwell-fts/src/main/scala/cmwell/fts/FTSService.scala [hunk 4]
```diff
@@ -540,7 +545,22 @@ class FTSService(config: Config) extends NsSplitter{
               logger.error(s"exhausted all retries attempts. logging failures to RED_LOG and returning results ")
               promise.success(SuccessfulBulkIndexResult(indexRequests, bulkResponse))
             }
-          } else {
+          } else if(versionConflictErrors.length > 0){
+            val bulks = versionConflictErrors.map { sir =>
+              get(sir._1.getId, sir._1.getIndex)(executionContext).transform {
+                case Success(esInfotonRes) => esInfotonRes match {
+                  case Some(ti) => Success(SuccessfulIndexResult(sir._1.getId, Option(ti._1.indexTime)))
+                  case None => Success(SuccessfulIndexResult(sir._1.getId, None))
```
Comment (e-orz) [inline server/cmwell-fts/src/main/scala/cmwell/fts/FTSService.scala]: after version conflict the response should never be None!!!
print and logs and "convert" it to nonRecoverableBulkIndexResults

File: server/cmwell-fts/src/main/scala/cmwell/fts/FTSService.scala [hunk 5]
```diff
@@ -540,7 +545,22 @@ class FTSService(config: Config) extends NsSplitter{
...
+                }
+                case Failure(e) => logger.error("Got exception from es while get index time for version conflict, error=" +
+                  e.getMessage + ",uuid=" + sir._1.getId)
+                  Failure(e)
```
Comment (e-orz) [inline server/cmwell-fts/src/main/scala/cmwell/fts/FTSService.scala]: same as before, it should be returned as a failed one.

File: server/cmwell-fts/src/main/scala/cmwell/fts/FTSService.scala [hunk 6]
```diff
@@ -540,7 +545,22 @@ class FTSService(config: Config) extends NsSplitter{
...
-          } else {
+          } else if(versionConflictErrors.length > 0){
+            val bulks = versionConflictErrors.map { sir =>
+              get(sir._1.getId, sir._1.getIndex)(executionContext).transform {
+                case Success(esInfotonRes) => esInfotonRes match {
+                  case Some(ti) => Success(SuccessfulIndexResult(sir._1.getId, Option(ti._1.indexTime)))
+                  case None => Success(SuccessfulIndexResult(sir._1.getId, None))
+                }
+                case Failure(e) => logger.error("Got exception from es while get index time for version conflict, error=" +
+                  e.getMessage + ",uuid=" + sir._1.getId)
```
Comment (e-orz) [inline server/cmwell-fts/src/main/scala/cmwell/fts/FTSService.scala]: always print stack trace of exceptions.

**Review by e-orz — CHANGES_REQUESTED** [review]

File: server/cmwell-fts/src/main/scala/cmwell/fts/FTSService.scala (hunk 1 above)
Comment (lielShraga) [reply server/cmwell-fts/src/main/scala/cmwell/fts/FTSService.scala]: done

File: server/cmwell-fts/src/main/scala/cmwell/fts/FTSService.scala (hunk 2 above)
Comment (lielShraga) [reply server/cmwell-fts/src/main/scala/cmwell/fts/FTSService.scala]: done

File: server/cmwell-fts/src/main/scala/cmwell/fts/FTSService.scala (hunk 3 above)
Comment (lielShraga) [reply server/cmwell-fts/src/main/scala/cmwell/fts/FTSService.scala]: done

Comment (lielShraga) [reply server/cmwell-fts/src/main/scala/cmwell/fts/FTSService.scala]: changed the impl..removed it.so done

File: server/cmwell-fts/src/main/scala/cmwell/fts/FTSService.scala (hunk 4 above)
Comment (lielShraga) [reply server/cmwell-fts/src/main/scala/cmwell/fts/FTSService.scala]: done

Comment (lielShraga) [reply server/cmwell-fts/src/main/scala/cmwell/fts/FTSService.scala]: I created a bulkResult and separate it to sucessfulList and failedList

File: server/cmwell-fts/src/main/scala/cmwell/fts/FTSService.scala (hunk 5 above)
Comment (lielShraga) [reply server/cmwell-fts/src/main/scala/cmwell/fts/FTSService.scala]: done. returned as part of failed bulk

File: server/cmwell-fts/src/main/scala/cmwell/fts/FTSService.scala (hunk 6 above)
Comment (lielShraga) [reply server/cmwell-fts/src/main/scala/cmwell/fts/FTSService.scala]: done

File: server/cmwell-fts/src/main/scala/cmwell/fts/FTSService.scala [hunk 7]
```diff
@@ -574,6 +587,33 @@ class FTSService(config: Config) extends NsSplitter{
...
+              bulkResponse._1.getItemId))
+        }
+        case Failure(e) =>
+          logger.error("Got exception from es while get index time in version conflict case, uuid=" + bulkResponse._1.getId, e)
+          Success(FailedIndexResult(bulkResponse._1.getId, e.getMessage, bulkResponse._1.getItemId))
+      }
+    }
+    val allIndexes = cmwell.util.concurrent.successes(bulks.toList)
+    val successAndFailedIndexes = allIndexes.map(indexResultList => {
+      val successfulIndexResults = indexResultList.collect { case index: SuccessfulIndexResult => index }
```
Comment (e-orz) [inline server/cmwell-fts/src/main/scala/cmwell/fts/FTSService.scala:601]: You should use match instead (2N complexity, you loose the compiler check of sealed trait match)

File: server/cmwell-fts/src/main/scala/cmwell/fts/FTSService.scala [hunk 8]
```diff
@@ -533,14 +539,21 @@ class FTSService(config: Config) extends NsSplitter{
...
+              val vConflictAndreRes = Future.sequence(List(reResponse, versionConflictBulkIndexResults)).
+                map(res => res.fold(SuccessfulBulkIndexResult(Nil, Nil))(_ ++ _))
               promise.completeWith(
-                reResponse.map { bulkSuccessfulResult ++ nonRecoverableBulkIndexResults ++ _ }
+                vConflictAndreRes.map {bulkSuccessfulResult ++ nonRecoverableBulkIndexResults ++ _}
               )
             } else {
               logger.error(s"exhausted all retries attempts. logging failures to RED_LOG and returning results ")
-              promise.success(SuccessfulBulkIndexResult(indexRequests, bulkResponse))
+              promise.completeWith(versionConflictBulkIndexResults.map{SuccessfulBulkIndexResult(indexRequests, bulkResponse) ++ _})
```
Comment (e-orz) [inline server/cmwell-fts/src/main/scala/cmwell/fts/FTSService.scala]: `indexRequests` contains also the `versionConflict` ones.

**Review by e-orz — APPROVED** [review]
Well done!

**Pull request merged**